src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Tooltip/DefaultTagHelperTooltipFactory.cs

```diff
@@ -381,5 +403,252 @@ private MarkupKind GetMarkupKind()
+            {
+                // There are certain characters that should be classified as plain text. For example,
+                // in 'TypeName<T, T2>', the characters '<', ',' and '>' should be classified as plain
+                // text while the rest should be classified as a keyword or type.
+                if (ch == '<' || ch == '>' || ch == '[' || ch == ']' || ch == ',')
+                {
+                    if (currentRunText.Length != 0)
+                    {
+                        var currentRunTextStr = currentRunText.ToString();
+
+                        // The type we're working with could contain a nested type, in which case we may
+                        // also need to reduce the inner type name(s), e.g. 'List<NamespaceName.TypeName>'
+                        if ((ch == '<' || ch == '>') && currentRunTextStr.Contains("."))
+                        {
+                            var reducedName = ReduceTypeName(currentRunText.ToString());
+                            ClassifyTextAsKeywordOrType(runs, reducedName);
+                        }
+                        else
+                        {
+                            ClassifyTextAsKeywordOrType(runs, currentRunText.ToString());
+                        }
+
+                        currentRunText.Clear();
+                    }
+
+                    runs.Add(new RazorClassifiedTextRun(RazorPredefinedClassificationTypeNames.Punctuation, ch.ToString()));
+                }
+                else
+                {
+                    currentRunText.Append(ch);
+                }
+            }
+
+            if (currentRunText.Length != 0)
+            {
+                ClassifyTextAsKeywordOrType(runs, currentRunText.ToString());
+            }
+        }
+
+        private static void ClassifyTextAsKeywordOrType(List<RazorClassifiedTextRun> runs, string typeName)
+        {
+            var nullableType = typeName.EndsWith("?");
+            if (nullableType)
+            {
+                // Classify the '?' symbol separately from the rest of the type since it's considered punctuation.
+                typeName = typeName.Substring(0, typeName.Length - 1);
+            }
+
+            // Case 1: Type can be aliased as a C# built-in type (e.g. Boolean -> bool, Int32 -> int, etc.).
+            if (TypeNameToAlias.TryGetValue(typeName, out var aliasedTypeName))
+            {
+                runs.Add(new RazorClassifiedTextRun(RazorPredefinedClassificationTypeNames.Keyword, aliasedTypeName));
+            }
+            // Case 2: Type is a C# built-in type (e.g. bool, int, etc.).
+            else if (CSharpBuiltInTypes.Contains(typeName))
+            {
+                runs.Add(new RazorClassifiedTextRun(RazorPredefinedClassificationTypeNames.Keyword, typeName));
+            }
+            // Case 3: All other types.
+            else
+            {
+                var typeNameParts = typeName.Split('.');
```

Comment:
With how you have it all setup this will always be a 1 length array wont it? Aka I wouldn't imagine typenames at this level would ever have dots in them